### PR TITLE
Fix the memory space for gradients when parameters are offloaded.

### PR DESCRIPTION
**Pull request by zhenying-liu**
When parameter_memory_host_offload=True in MaxText, parameters carry a pinned_host memory space in their JAX abstract values. JAX's AD machinery propagates this host memory space to gradients via to_tangent_aval / to_ct_aval where both functions pass memory_space=self.memory_space to the returned ShapedArray. It causes a ValueError when gradient clipping computes the global norm. The add in jnp.sqrt(sum(g\*\*2 for g in grads)) receives a mix of pinned_host gradients (from host-offloaded parameters) and device gradients, resulting in mismatched memory spaces.

This commit forces to return the ShapedArray with the explicit device memory space.

### CODE REVIEW

File: tests/core_test.py [hunk 1]
```diff
@@ -369,6 +369,18 @@ def test_aval_str_short_mem_space(self):
                             memory_space=jax.memory.Space.Device)
     self.assertEqual(aval.str_short(True), "f32[8]")
 
+  def test_grad_memory_space_is_device(self):
+    # Gradients should always be on device, even when parameters carry a
```
Comment (yashk2810) [inline tests/core_test.py:373]: I don't think this makes sense. What problem are you trying to solve?

Why shouldn't to_ct_aval and to_tangent_aval propagate the memory spaces?

Comment (zhenying-liu) [reply tests/core_test.py:373]: The gradient will be propagated host memory space when the parameters are offloaded. We don't want the gradient to be on the host memory space.

Comment (yashk2810) [reply tests/core_test.py]: Who's "we" here?

> We don't want the gradient to be on the host memory space.

Why not?

Comment (yashk2810) [reply tests/core_test.py:373]: Is there a small repro we can come up with? Then we can try to fix the problem.

Comment (yashk2810) [reply tests/core_test.py:373]: Another thing we can do is remove the check in bwd_pass that I added to make sure memories match up between primals and cotangets. Which I assume is causing a problem here?

If you have a stack trace or more information that would help figure out what's going wrong here.

Comment (zhenying-liu) [reply tests/core_test.py:373]: In the host memory offloading, a tensor gets the memory space through NamedSharding. The memory space of the gradients can't be decided through NamedSharding. If so, I think it should have the default memory space.

Comment (yashk2810) [reply tests/core_test.py:373]: Can you paste the stack trace of the error you got? And if you have a small repro that would be helpful!
